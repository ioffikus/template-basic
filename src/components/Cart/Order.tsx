import React from 'react';
import { connect } from 'react-redux';
import { Row, Col, Button } from 'antd';
import _ from 'lodash';
import CartItems from 'src/components/CartItems';
import Price from 'src/components/Price';

interface IOwnProps {
  nextStep: () => void;
}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  totalPrice: number;
}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
  totalPrice: state.cart.totalPrice,
});

const WrappedOrder = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Order extends React.Component<IProps, IState> {
    render() {
      // TODO ch1333
      const currency: string = _.get(this.props, 'items[0].schema_fields.currency.value', '');
      const buttonsUI = this.props.ui.common.buttons;
      return (
        <div className="root">
          <style jsx>{`
            .cart-menu-wrapper {
              margin-bottom: 25px;
            }
            .button-wrapper {
              text-align: right;
            }
            .total {
              margin-bottom: 30px;
            }
          `}</style>
          <div className="cart-menu-wrapper">
            <CartItems
              cssModificationClass="has-border-bottom"
              mayDeleteItems={true}
              mayChangeAmountItems={true}
              size="large"
            />
          </div>
          <Row type="flex" justify="end" align="middle">
            <Col span={23} pull={1}>
              <div className="total">
                {this.props.ui.routes.cart.total.i18n.defaultText}:
                <span className="price-wrapper">
                  <Price price={this.props.totalPrice} currency={currency} />
                </span>
              </div>
            </Col>
          </Row>
          <div className="button-wrapper">
            <Button type="primary" size="large" onClick={this.props.nextStep}>
              {buttonsUI.nextStep.i18n.defaultText}
            </Button>
          </div>
        </div>
      );
    }
  },
);

export default WrappedOrder;
