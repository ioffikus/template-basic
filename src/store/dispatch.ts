import { Store } from 'redux';
import _ from 'lodash';
import { showError } from 'src/store/common';

let store: Store<Store.IState>;

export const setupDispatch = (config: { store: Store<Store.IState> }) => {
  store = config.store;
};

const dispatch = (action: any) => {
  if (_.isFunction(action)) {
    return store.dispatch(action).catch((error: any) => {
      if (!_.isNull(error)) {
        showError(error);
      }
      return Promise.reject(error);
    });
  }
  return store.dispatch(action);
};

export default dispatch;
