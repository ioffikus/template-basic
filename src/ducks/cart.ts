import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import _ from 'lodash';
import { BusinessLogicError } from 'src/store/common';
import { safelyParseJSON } from 'src/helpers/safelyParseJSON';
import { IObject } from 'src/core/interfaces/IObject';
import { GET, POST, PUT, DELETE } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CART';
const NS = `${REDUCER}__`;

export interface ICartState extends Store.IReduxCommonState {
  id: number;
  items: Alicanto.Models.CatalogItem[];
  itemsCount: number;
  totalPrice: number;
}

export const initialState: ICartState = {
  id: null,
  items: [],
  itemsCount: 0,
  totalPrice: 0,
  isRequestError: false,
  isRequesting: false,
};

const reducer = createReducer<ICartState>({}, initialState);

const addProductToCart = createAction<Alicanto.Models.CatalogItem>(`${NS}ADD_PRODUCT_TO_CART`);
reducer.on(addProductToCart, (state, product) => ({
  ...state,
  items: [
    ...state.items,
    {
      ...product,
    },
  ],
  itemsCount: state.itemsCount + 1,
  totalPrice: (state.totalPrice * 100 + product.schema_fields.price.value * 100) / 100,
}));

const deleteProductFromCart = createAction<number>(`${NS}DELETE_PRODUCT_FROM_CART`);
reducer.on(deleteProductFromCart, (state, index) => ({
  ...state,
  items: state.items.filter((_, i) => i !== index),
  itemsCount: state.itemsCount - state.items[index].count,
  totalPrice: (state.totalPrice * 100 - state.items[index].schema_fields.price.value * 100) / 100,
}));

const increaseProductQuantity = createAction<number>(`${NS}INCREASE_PRODUCT_QUANTITY`);
reducer.on(increaseProductQuantity, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        count: state.items[index].count + 1,
      },
    ])
    .concat(state.items.slice(index + 1)),
  itemsCount: state.itemsCount + 1,
  totalPrice: (state.totalPrice * 100 + state.items[index].schema_fields.price.value * 100) / 100,
}));

const decreaseProductQuantity = createAction<number>(`${NS}DECREASE_PRODUCT_QUANTITY`);
reducer.on(decreaseProductQuantity, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        count: state.items[index].count - 1,
      },
    ])
    .concat(state.items.slice(index + 1)),
  itemsCount: state.itemsCount - 1,
  totalPrice: (state.totalPrice * 100 - state.items[index].schema_fields.price.value * 100) / 100,
}));

const blockItemIncrease = createAction<number>(`${NS}BLOCK_ITEM_INCREASE`);
reducer.on(blockItemIncrease, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        blockedIncreasing: true,
      },
    ])
    .concat(state.items.slice(index + 1)),
}));

const putProductsInStore = createAction<Alicanto.Models.CatalogItem[]>(`${NS}PUT_PRODUCTS_IN_STORE`);
reducer.on(putProductsInStore, (state, items) => ({
  ...state,
  items: [...items],
}));

const putCartInStore = createAction<{ id: number; itemsCount: number; totalPrice: number }>(`${NS}PUT_CART_IN_STORE`);
reducer.on(putCartInStore, (state, { id, itemsCount, totalPrice }) => ({
  ...state,
  id,
  itemsCount,
  totalPrice,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  isRequestError: true,
  isRequesting: false,
}));

const callSuccess = createAction(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, state => ({
  ...state,
  isRequestError: false,
  isRequesting: false,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequestError: false,
  isRequesting: true,
}));

export const getCartFromLocalStorage = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  if (process.env.BROWSER) {
    const s = window.localStorage.getItem('cart');
    if (s) {
      const localStorageCartState: ICartState = safelyParseJSON(s);
      if (!_.isNull(localStorageCartState)) {
        const { id, itemsCount, totalPrice, items } = localStorageCartState;
        dispatch(putCartInStore({ id, itemsCount, totalPrice }));
        dispatch(putProductsInStore(items));
      }
    }
  }
  return Promise.resolve();
};

const setCartInLocalStorage = (cartState: ICartState) => {
  if (process.env.BROWSER) {
    const { items, itemsCount, totalPrice, id } = cartState;
    window.localStorage.setItem('cart', JSON.stringify({ items, itemsCount, totalPrice, id }));
  }
};

const removeCartFromLocalStorage = () => {
  if (process.env.BROWSER) {
    window.localStorage.removeItem('cart');
  }
};

const getProducts = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { results } = resp.data as Alicanto.API.GET_CART_CURRENT_ITEMS_RESPONSE;
        dispatch(putProductsInStore(results));
        dispatch(callSuccess());
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

const getCart = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { items_count, total, id } = resp.data as Alicanto.API.GET_CART_CURRENT_RESPONSE;
        dispatch(putCartInStore({ id, itemsCount: items_count, totalPrice: parseFloat(total) }));
        dispatch(callSuccess());
        if (!items_count) {
          dispatch(deleteCart());
          resolve();
          return;
        }
        return dispatch(getProducts()).then(() => {
          setCartInLocalStorage(getState().cart);
          resolve();
        });
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

/**
 * If unauthenticated user've added some products, then
 * after login we put items on server,
 * firstly check if there are repeated items
 * in local storage and server. After that update cart info
 */
const putCart = (data: Alicanto.API.PUT_CART_CURRENT_BODY): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    PUT(
      {
        data,
        url: `${EndpointsConfig.CART_CURRENT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(callSuccess());
        return dispatch(getCart()).then(resolve);
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

const findUnicItems = (itemsInStore: Alicanto.Models.CatalogItem[], itemsInCart: Alicanto.Models.CatalogItem[]) => {
  const cartIds: IObject<Alicanto.Models.CatalogItem> = {};
  itemsInCart.forEach((item: Alicanto.Models.CatalogItem) => {
    cartIds[item.item] = item;
  });
  return itemsInStore.filter(item => !(item.item in cartIds));
};

export const postLocalStorageCart = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    if (_.isEmpty(items)) {
      dispatch(getCart()).then(resolve, reject);
      return;
    }
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { results } = resp.data as Alicanto.API.GET_CART_CURRENT_ITEMS_RESPONSE;
        const itemsInCart = results;
        const unicItemsInStore = findUnicItems(items, itemsInCart);
        const data: Alicanto.API.PUT_CART_CURRENT_BODY = {
          items: [],
        };
        if (!_.isEmpty(unicItemsInStore)) {
          unicItemsInStore.forEach(item => {
            data.items.push({
              count: item.count,
              item: item.item,
            });
          });
          return dispatch(putCart(data)).then(resolve);
        }
        dispatch(callSuccess());
        return dispatch(getCart()).then(resolve);
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

/**
 * Dispatch action after click on 'add to card' button:
 * first click will add item in the cart
 * user can press the button multiple times and this action will increase product quantity
 * @param product
 */
export const addToCart = (product: Alicanto.Models.CatalogItem, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const productBalance = product.balance;
    const productIsUnlimited = product.is_unlimited;
    const productUid = product.uid;
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    const errorMsg = getState().ui.routes.cart.errors.moreThanInBalance.i18n.defaultText;
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      if (itemIndex !== -1) {
        const item = items[itemIndex];
        if (productBalance < item.count + 1 && !productIsUnlimited) {
          dispatch(callFailure());
          const error = new BusinessLogicError(errorMsg);
          reject(error);
          return;
        }
        const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
          count: item.count + 1,
          item: productUid,
        };
        PUT(
          {
            data,
            url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
          },
          xShopHeaders,
        )
          .then((resp: AxiosResponse) => {
            dispatch(increaseProductQuantity(itemIndex));
            setCartInLocalStorage(getState().cart);
            dispatch(callSuccess());
            resolve();
          })
          .catch(error => {
            dispatch(callFailure());
            reject(error);
          });
        return;
      }
      const data: Alicanto.API.POST_CART_CURRENT_ITEMS_BODY = {
        count: 1,
        item: productUid,
      };
      POST(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          const data = resp.data as Alicanto.API.POST_CART_CURRENT_ITEMS_RESPONSE;
          dispatch(addProductToCart(data));
          if (cb) {
            cb();
          }
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    if (itemIndex !== -1) {
      dispatch(callRequest());
      GET(
        {
          url: `${EndpointsConfig.PRODUCT}${product.uid}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
          const item = items[itemIndex];
          if (item.count + 1 < data.balance || data.is_unlimited) {
            dispatch(increaseProductQuantity(itemIndex));
            if (cb) {
              cb();
            }
            setCartInLocalStorage(getState().cart);
            dispatch(callSuccess());
            resolve();
            return;
          }
          const error = new BusinessLogicError(errorMsg);
          return Promise.reject(error);
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    const newProduct = { ...product };
    newProduct.count = 1;
    newProduct.item = product.uid;
    delete newProduct.uid;
    dispatch(addProductToCart(newProduct));
    if (cb) {
      cb();
    }
    setCartInLocalStorage(getState().cart);
    resolve();
  });
};

export const deleteFromCart = (productUid: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      DELETE(
        {
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(deleteProductFromCart(itemIndex));
          if (getState().cart.itemsCount === 0) {
            dispatch(deleteCart());
          } else {
            setCartInLocalStorage(getState().cart);
          }
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    dispatch(deleteProductFromCart(itemIndex));
    if (getState().cart.itemsCount === 0) {
      dispatch(deleteCart());
    } else {
      setCartInLocalStorage(getState().cart);
    }
    resolve();
  });
};

export const increaseQuantity = (productUid: string, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    if (itemIndex === -1) {
      resolve();
      return;
    }
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    dispatch(callRequest());
    if (isAuthenticated) {
      const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
        count: item.count + 1,
        item: productUid,
      };
      PUT(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(increaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(blockItemIncrease(itemIndex));
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    GET(
      {
        url: `${EndpointsConfig.PRODUCT}${productUid}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
        if (item.count + 1 < data.balance || data.is_unlimited) {
          dispatch(increaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
          return;
        }
        const errorMsg = getState().ui.routes.cart.errors.moreThanInBalance.i18n.defaultText;
        return Promise.reject(new BusinessLogicError(errorMsg));
      })
      .catch(error => {
        dispatch(blockItemIncrease(itemIndex));
        dispatch(callFailure());
        reject(error);
      });
  })
    .then(() => {
      if (cb) {
        cb();
      }
    })
    .catch(error => {
      if (cb) {
        cb();
      }
      return Promise.reject(error);
    });
};

export const decreaseQuantity = (productUid: string, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    if (itemIndex === -1) {
      resolve();
      return;
    }
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
        count: item.count - 1,
        item: productUid,
      };
      PUT(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(decreaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    if (item.count > 1) {
      dispatch(decreaseProductQuantity(itemIndex));
      setCartInLocalStorage(getState().cart);
      resolve();
      return;
    }
    const errorMsg = getState().ui.routes.cart.errors.fewerThanOne.i18n.defaultText;
    const error = new BusinessLogicError(errorMsg);
    reject(error);
  })
    .then(() => {
      if (cb) {
        cb();
      }
    })
    .catch(error => {
      if (cb) {
        cb();
      }
      return Promise.reject(error);
    });
};

export const deleteCart = createAction(`${NS}RESET`);
reducer.on(deleteCart, state => {
  removeCartFromLocalStorage();
  return {
    ...state,
    itemsCount: 0,
    totalPrice: 0,
    items: [],
  };
});

export default reducer;
