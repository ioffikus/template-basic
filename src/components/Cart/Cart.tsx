import React from 'react';
import { connect } from 'react-redux';
import { Row, Col, Button } from 'antd';
import _ from 'lodash';
import dispatch from 'src/store/dispatch';
import { checkoutUserOrder } from 'src/ducks/userOrders';
import CartItems from 'src/components/CartItems';
import Price from 'src/components/Price';

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  items: Alicanto.Models.CatalogItem[];
  totalPrice: number;
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  isMounted: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
  items: state.cart.items,
  totalPrice: state.cart.totalPrice,
});

const WrappedCart = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Cart extends React.Component<IProps, IState> {
    state: IState = {
      isMounted: false,
    };

    handleCheckout = () => {
      dispatch(checkoutUserOrder());
    };

    componentDidMount() {
      this.setState({ isMounted: true });
    }

    render() {
      // TODO ch1333
      const currency: string = _.get(this.props, 'items[0].schema_fields.currency.value', '');
      const cartUI = this.props.ui.routes.cart;
      return (
        <div className="root app-container">
          <style jsx>{`
            .root {
              padding-bottom: 30px;
            }
            .root :global(.total) {
              text-align: right;
              font-size: 10px;
              color: #b2bdcb;
            }
            .root :global(.price-wrapper) {
              margin-left: 5px;
              color: #062f49;
              font-size: 18px;
              font-weight: bold;
            }
            .cart-menu-wrapper {
              margin-bottom: 25px;
            }
            .button-wrapper {
              text-align: right;
            }
            .total {
              margin-bottom: 30px;
            }
          `}</style>
          {this.state.isMounted &&
            (_.isEmpty(this.props.items) ? (
              <h2 className="app-text-empty">{cartUI.emptyStatus.i18n.defaultText}</h2>
            ) : (
              <div>
                <div className="cart-menu-wrapper">
                  <CartItems
                    cssModificationClass="has-border-bottom"
                    mayDeleteItems={true}
                    mayChangeAmountItems={true}
                    size="large"
                  />
                </div>
                <Row type="flex" justify="end" align="middle">
                  <Col span={23} pull={1}>
                    <div className="total">
                      {this.props.ui.routes.cart.total.i18n.defaultText}:
                      <span className="price-wrapper">
                        <Price price={this.props.totalPrice} currency={currency} />
                      </span>
                    </div>
                  </Col>
                </Row>
                <div className="button-wrapper">
                  <Button type="primary" size="large" onClick={this.handleCheckout}>
                    {cartUI.checkout.i18n.defaultText}
                  </Button>
                </div>
              </div>
            ))}
        </div>
      );
    }
  },
);

export default WrappedCart;
