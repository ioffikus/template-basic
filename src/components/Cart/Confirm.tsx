import React from 'react';
import { connect } from 'react-redux';
import { Row, Col, Button } from 'antd';
import _ from 'lodash';
import dispatch from 'src/store/dispatch';
import { IUserDeliveryState } from 'src/ducks/userDelivery';
import { checkoutUserOrder } from 'src/ducks/userOrders';
import CartItems from 'src/components/CartItems';
import Price from 'src/components/Price';

interface IOwnProps {
  prevStep: () => void;
  nextStep: () => void;
}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  userDeliveryValues: IUserDeliveryState;
  totalPrice: number;
}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
  userDeliveryValues: state.userDelivery,
  totalPrice: state.cart.totalPrice,
});

const WrappedConfirm = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Confirm extends React.Component<IProps, IState> {
    handleConfirmClick = () => {
      dispatch(checkoutUserOrder());
    };

    render() {
      const confirmUI = this.props.ui.routes.cart.confirm;
      const orderInfo = [
        {
          title: confirmUI.labels.purchaser.i18n.defaultText,
          value: `${this.props.userDeliveryValues.name} ${this.props.userDeliveryValues.surname}`,
        },
        {
          title: confirmUI.labels.phoneNumber.i18n.defaultText,
          value: `+${this.props.userDeliveryValues.phonePrefix} ${this.props.userDeliveryValues.phoneNumber}`,
        },
        {
          title: confirmUI.labels.email.i18n.defaultText,
          value: this.props.userDeliveryValues.email,
        },
        {
          title: confirmUI.labels.address.i18n.defaultText,
          value: `${this.props.userDeliveryValues.address}, ${this.props.userDeliveryValues.stateRegion}, ${this.props
            .userDeliveryValues.country}, ${this.props.userDeliveryValues.zipCode}`,
        },
      ];
      const buttonsUI = this.props.ui.common.buttons;
      // TODO ch1333
      const currency = _.get(this.props, 'items[0].schema_fields.currency.value', '');
      return (
        <div className="root">
          <style jsx>{`
            .cart-menu-wrapper {
              margin-bottom: 17px;
            }
            .info {
              column-count: 2;
              column-gap: 10px;
            }
            .root :global(.info-wrapper) {
              margin-bottom: 28px;
            }
            .root :global(.item) {
              margin-bottom: 12px;
            }
            .title {
              color: #b2bdcb;
              font-size: 10px;
              font-weight: 500;
            }
            .value {
              font-weight: bold;
            }
          `}</style>
          <div className="cart-menu-wrapper">
            <CartItems cssModificationClass="has-border-bottom" size="large" />
          </div>
          <Row type="flex" justify="space-between" className="info-wrapper">
            <Col span={18}>
              <div className="info">
                {orderInfo.map(item => (
                  <Row key={item.title} type="flex" className="item">
                    <Col span={7}>
                      <div className="title">{item.title}:</div>
                    </Col>
                    <Col span={17}>
                      <div className="value">{item.value}</div>
                    </Col>
                  </Row>
                ))}
              </div>
            </Col>
            <Col span={6}>
              <div className="total">
                {this.props.ui.routes.cart.total.i18n.defaultText}:
                <span className="price-wrapper">
                  <Price price={this.props.totalPrice} currency={currency} />
                </span>
              </div>
            </Col>
          </Row>
          <Row type="flex" justify="space-between">
            <Col>
              <Button type="primary" size="large" ghost onClick={this.props.prevStep}>
                {buttonsUI.prevStep.i18n.defaultText}
              </Button>
            </Col>
            <Col>
              <Button type="primary" size="large" onClick={this.props.nextStep}>
                {buttonsUI.confirm.i18n.defaultText}
              </Button>
            </Col>
          </Row>
        </div>
      );
    }
  },
);

export default WrappedConfirm;
