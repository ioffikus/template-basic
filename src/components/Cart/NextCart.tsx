import React from 'react';
import { connect } from 'react-redux';
import { Tabs } from 'antd';
import _ from 'lodash';
import { IObject } from 'src/core/interfaces/IObject';
import Delivery from 'src/components/CartDelivery';
import Payment from 'src/components/CartPayment';
import Order from './Order';
import Confirm from './Confirm';

// WIP: It is future version of Cart

interface ITabProps {
  prevStep: () => void;
  nextStep: () => void;
}

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  items: Alicanto.Models.CatalogItem[];
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  activeKey: string;
  isMounted: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
  items: state.cart.items,
});

const WrappedCart = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Cart extends React.Component<IProps, IState> {
    readonly defaultKey = 'order';

    readonly tabs: IObject<React.StatelessComponent<ITabProps>> = {
      order: props => <Order {...props} />,
      delivery: props => <Delivery {...props} />,
      confirm: props => <Confirm {...props} />,
      payment: props => <Payment {...props} />,
    };

    readonly tabKeys = Object.keys(this.tabs);

    state: IState = {
      activeKey: this.defaultKey,
      isMounted: false,
    };

    nextStep = (key: string) => () => {
      const index = this.tabKeys.indexOf(key);
      this.setState({ activeKey: this.tabKeys[index + 1] });
    };

    prevStep = (key: string) => () => {
      const index = this.tabKeys.indexOf(key);
      this.setState({ activeKey: this.tabKeys[index - 1] });
    };

    onTabClick = (activeKey: string) => {
      this.setState({ activeKey });
    };

    componentDidMount() {
      this.setState({ isMounted: true });
    }

    render() {
      const cartUI = this.props.ui.routes.cart;
      return (
        <div className="root app-container">
          <style jsx>{`
            .root {
              padding-bottom: 30px;
            }
            .root :global(.ant-tabs-bar) {
              margin-bottom: 29px;
            }
            .root :global(.total) {
              text-align: right;
              font-size: 10px;
              color: #b2bdcb;
            }
            .root :global(.price-wrapper) {
              margin-left: 5px;
              color: #062f49;
              font-size: 18px;
              font-weight: bold;
            }
          `}</style>
          {this.state.isMounted &&
            (_.isEmpty(this.props.items) ? (
              <h2 className="app-text-empty">{cartUI.emptyStatus.i18n.defaultText}</h2>
            ) : (
              <Tabs activeKey={this.state.activeKey} onTabClick={this.onTabClick}>
                {this.tabKeys.map(key => {
                  const title = _.get(cartUI.tabs, `${key}.i18n.defaultText`, '');
                  const Tab = this.tabs[key];
                  return (
                    <Tabs.TabPane
                      key={key}
                      tab={title}
                      disabled={this.tabKeys.indexOf(key) >= this.tabKeys.indexOf(this.state.activeKey)}
                    >
                      <Tab prevStep={this.prevStep(key)} nextStep={this.nextStep(key)} />
                    </Tabs.TabPane>
                  );
                })}
              </Tabs>
            ))}
        </div>
      );
    }
  },
);

export default WrappedCart;
