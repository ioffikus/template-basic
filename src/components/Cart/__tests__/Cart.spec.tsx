import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { Button } from 'antd';
import configureMockStore from 'redux-mock-store';
import { setupDispatch } from 'src/store/dispatch';
import { checkoutUserOrder } from 'src/ducks/userOrders';
import uiStateMock from 'src/store/__mocks__/uiStore';
import cartStateMock from 'src/store/__mocks__/cartStore';
import { initialState as cartInitialState } from 'src/ducks/cart';
import Cart from '../Cart';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('src/ducks/userOrders', () => {
  return {
    checkoutUserOrder: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});

describe(null, () => {
  it('render() with empty items', () => {
    const store = mockStore({ ui: uiStateMock, cart: cartInitialState });
    const wrapper = shallow(<Cart />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with items', () => {
    const store = mockStore({ ui: uiStateMock, cart: cartStateMock });
    const wrapper = shallow(<Cart />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('Button.onClick() to Button', () => {
    const store = mockStore({ ui: uiStateMock, cart: cartStateMock });
    setupDispatch({ store });
    const wrapper = shallow(<Cart />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const buttonInstance = dummyWrapper
      .find(Button)
      .dive()
      .instance() as Button;
    buttonInstance.props.onClick(null);
    expect(checkoutUserOrder).toBeCalled();
  });
});
