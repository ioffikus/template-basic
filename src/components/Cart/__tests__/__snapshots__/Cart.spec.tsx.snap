// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() with empty items 1`] = `
<div
  className="jsx-265714823 root app-container"
>
  <JSXStyle
    css=".root.jsx-265714823{padding-bottom:30px;}.root.jsx-265714823 .total{text-align:right;font-size:10px;color:#b2bdcb;}.root.jsx-265714823 .price-wrapper{margin-left:5px;color:#062f49;font-size:18px;font-weight:bold;}.cart-menu-wrapper.jsx-265714823{margin-bottom:25px;}.button-wrapper.jsx-265714823{text-align:right;}.total.jsx-265714823{margin-bottom:30px;}"
    styleId="265714823"
  />
  <h2
    className="jsx-265714823 app-text-empty"
  >
    Please add a product to card
  </h2>
</div>
`;

exports[`render() with items 1`] = `
<div
  className="jsx-265714823 root app-container"
>
  <JSXStyle
    css=".root.jsx-265714823{padding-bottom:30px;}.root.jsx-265714823 .total{text-align:right;font-size:10px;color:#b2bdcb;}.root.jsx-265714823 .price-wrapper{margin-left:5px;color:#062f49;font-size:18px;font-weight:bold;}.cart-menu-wrapper.jsx-265714823{margin-bottom:25px;}.button-wrapper.jsx-265714823{text-align:right;}.total.jsx-265714823{margin-bottom:30px;}"
    styleId="265714823"
  />
  <div
    className="jsx-265714823"
  >
    <div
      className="jsx-265714823 cart-menu-wrapper"
    >
      <Connect(CartItems)
        cssModificationClass="has-border-bottom"
        mayChangeAmountItems={true}
        mayDeleteItems={true}
        size="large"
      />
    </div>
    <Row
      align="middle"
      gutter={0}
      justify="end"
      type="flex"
    >
      <Col
        pull={1}
        span={23}
      >
        <div
          className="jsx-265714823 total"
        >
          Total
          :
          <span
            className="jsx-265714823 price-wrapper"
          >
            <Connect(Price)
              currency="â‚¬"
              price={25}
            />
          </span>
        </div>
      </Col>
    </Row>
    <div
      className="jsx-265714823 button-wrapper"
    >
      <Button
        ghost={false}
        loading={false}
        onClick={[Function]}
        prefixCls="ant-btn"
        size="large"
        type="primary"
      >
        Check Out
      </Button>
    </div>
  </div>
</div>
`;
